sky/clouds/ibm.py

```diff
@@ -0,0 +1,502 @@
+
+    def accelerators_to_hourly_cost(self,
+                                    accelerators: Dict[str, int],
+                                    use_spot: bool,
+                                    region: Optional[str] = None,
+                                    zone: Optional[str] = None) -> float:
+        del accelerators, use_spot, region, zone  # unused
+        # Currently Isn't implemented in the same manner by aws and azure.
+        return 0
+
+    def get_egress_cost(self, num_gigabytes):
+        """Returns the egress cost. Currently true for us-south, i.e. Dallas.
+        based on https://cloud.ibm.com/objectstorage/create#pricing. """
+        cost = 0
+        price_thresholds = [{
+            'threshold': 150,
+            'price_per_gb': 0.05
+        }, {
+            'threshold': 50,
+            'price_per_gb': 0.07
+        }, {
+            'threshold': 0,
+            'price_per_gb': 0.09
+        }]
+        for price_threshold in price_thresholds:
+            # pylint: disable=line-too-long
+            cost += (num_gigabytes - price_threshold['threshold']
+                    ) * price_threshold['price_per_gb']
+            num_gigabytes -= (num_gigabytes - price_threshold['threshold'])
+        return cost
+
+    def is_same_cloud(self, other):
+        return isinstance(other, IBM)
+
+    def make_deploy_resources_variables(
+        self,
+        resources: 'resources_lib.Resources',
+        region: Optional['clouds.Region'],
+        zones: Optional[List['clouds.Zone']],
+    ) -> Dict[str, Optional[str]]:
+        """Converts planned sky.Resources to cloud-specific resource variables.
+
+        These variables are used to fill the node type section (instance type,
+        any accelerators, etc.) in the cloud's deployment YAML template.
+
+        Cloud-agnostic sections (e.g., commands to run) need not be returned by
+        this function.
+
+        Returns:
+          A dictionary of cloud-specific node type variables.
+        """
+
+        def _get_profile_resources(instance_profile):
+            """returns a dict representing the
+             cpu, memory and gpu of specified instance profile"""
+            profile_resources_str = instance_profile.split('-')[1]
+            # gpu count based on profile
+            gpu_num = int(
+                profile_resources_str.split('x')[2].split('v')[0]) if (len(
+                    profile_resources_str.split('x')) == 3) else 0
+            # cpu count based on profile
+            cpu_num = int(profile_resources_str.split('x')[0])
+            # memory GBs on profile
+            memory_num = int(profile_resources_str.split('x')[1])
+            return {'CPU': cpu_num, 'memory': memory_num, 'GPU': gpu_num}
+
+        if region is None:
+            assert zones is None, (
+                'Set either both or neither for: region, zones.')
+            region = self._get_default_region()
+            zones = region.zones
```

Comment:
This will never be true, as we have now guaranteed that the region will be filled for the `resources` reaching here.

Reply:
Yes, i see the change you've made. From the typings of this [method](https://github.com/skypilot-org/skypilot/blob/559f9725905cdc1215ebc397384c362b7b55fb5d/sky/clouds/cloud.py#L221) I deduce that zones could potentially be None. I have to initialize zones in that case (zones is None while region isn't) , right?

Reply:
I should have made the docstr more clear, but the zones will not be None in this function for the clouds that support availability zones, as the zones are always passed in with the `IBM.zones_provision_loop`, in the following: https://github.com/skypilot-org/skypilot/blob/508be85e4a2a0df1fa98f7fac63980ff6e24da3e/sky/backends/cloud_vm_ray_backend.py#L1075-L1095